peer/record.go

```diff
@@ -0,0 +1,221 @@
+package peer
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/peer/pb"
+	"github.com/libp2p/go-libp2p-core/record"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+func init() {
+	record.RegisterPayloadType(PeerRecordEnvelopePayloadType, &PeerRecord{})
+}
+
+// The domain string used for peer records contained in a Envelope.
+const PeerRecordEnvelopeDomain = "libp2p-peer-record"
+
+// The type hint used to identify peer records in a Envelope.
+// TODO: register multicodec
+var PeerRecordEnvelopePayloadType = []byte("/libp2p/peer-record")
```

Comment:
We now have codec `0x0301` allocated for peer records: https://github.com/multiformats/multicodec/pull/157